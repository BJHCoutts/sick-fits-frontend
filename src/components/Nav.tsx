import Link from 'next/link'
import SNav from './styles/SNav'
import { useUser } from '../lib/functions/useUser'

export default function Nav() {
	const user = useUser()

	return (
		<SNav>
			<ul>
				<li>
					<Link href="/products">Products</Link>
				</li>
				{user ? (
					<>
						<li>
							<Link href="/sell">Sell</Link>
						</li>
						<li>
							<Link href="/orders">Orders</Link>
						</li>
						<li>
							<Link href="/account">Account</Link>
						</li>
					</>
				) : (
					<>
						<li>
							<Link href="/sign-in">Sign In</Link>
						</li>
					</>
				)}
			</ul>
		</SNav>
	)
}
