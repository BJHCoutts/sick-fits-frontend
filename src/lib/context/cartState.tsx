import { createContext, useContext, useState } from 'react'

const LocalStateContext = createContext(null)
const LocalStateProvider = LocalStateContext.Provider

function CartStateProvider({ children }) {
	const [cartOpen, setCartOpen] = useState(true)
	function toggleCart() {
		setCartOpen(!cartOpen)
	}
	function closeCart() {
		setCartOpen(false)
	}
	function openCart() {
		setCartOpen(true)
	}

	return (
		<LocalStateProvider value={{ cartOpen, toggleCart, closeCart, openCart }}>
			{children}
		</LocalStateProvider>
	)
}

function useCart() {
	// this is the consumer to access state
	const all = useContext(LocalStateContext)
	return all
}

export { CartStateProvider, useCart }
