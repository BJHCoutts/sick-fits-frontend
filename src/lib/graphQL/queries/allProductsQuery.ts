import { gql } from "@apollo/client"

export const ALL_PRODUCTS_QUERY = gql`
query ALL_PRODUCTS_QUERY{
  allProducts {
    id
    name
    description
    photo {
      id
      image {
        id
        publicUrlTransformed
      }
      altText
    }
    status
  }
}
`