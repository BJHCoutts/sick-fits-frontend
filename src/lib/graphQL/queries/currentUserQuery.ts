import { gql } from "@apollo/client"

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    authenticatedItem {
      ... on User {
        id
        email
        name
        cart {
          id
          quantity
          product {
            id
            name
            description
            price
            photo {
              id
              altText
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`