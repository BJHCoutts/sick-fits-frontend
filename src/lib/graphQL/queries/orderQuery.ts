import { gql } from "@apollo/client"

export const ORDER_QUERY = gql`
  query ORDER_QUERY($id: ID!) {
    order: Order(where: { id: $id }) {
      id
      charge
      total
      user {
        id
      }
      items {
        id
        name
        description
        price
        quantity
        photo {
          image {
            publicUrlTransformed
          }
        }
      }
    }
  }
`