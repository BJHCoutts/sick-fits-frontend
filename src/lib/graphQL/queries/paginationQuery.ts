import { gql } from "@apollo/client"

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY{
		_allProductsMeta {
			count
		}
	}
`