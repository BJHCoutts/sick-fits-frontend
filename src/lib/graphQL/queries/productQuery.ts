import { gql } from "@apollo/client"

export const PRODUCT_QUERY = gql`
	query PRODUCT_QUERY (
		$id: ID!
		){
		Product (where: {id: $id}) {
			id
			name
			description
			photo {
				id
				image {
					id
					publicUrlTransformed
				}
				altText
			}
			status
			price
		}
	}
`