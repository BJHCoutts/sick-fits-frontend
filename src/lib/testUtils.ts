import casual from 'casual'
import { PAGINATION_QUERY } from './graphQL/queries/paginationQuery'
import { TProduct } from './types/TProduct'
import { TCartItem } from './types/TCartItem'
import { TUser } from './types/TUser'

// seed it so we get consistent results
casual.seed(777)

export const fakeItem = (): TProduct => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	photo: {
		id: 'abc123',
		altText: 'dogs are best',
		image: {
			id: 'abc123',
			publicUrlTransformed: 'dog.jpg',
		},
	},
	name: 'dogs are best',
	description: 'dogs',
	status: 'AVAILABLE',
})

export const fakeUser = (overrides: {}): TUser => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...overrides,
})

// const fakeOrderItem = () => ({
//   __typename: 'OrderItem',
//   id: casual.uuid,
//   image: {
//     image: `${casual.word}.jpg`,
//   },
//   name: casual.words(),
//   price: 4234,
//   quantity: 1,
//   description: casual.words(),
// });

// const fakeOrder = () => ({
//   __typename: 'Order',
//   id: 'ord123',
//   charge: 'ch_123',
//   total: 40000,
//   items: [fakeOrderItem(), fakeOrderItem()],
//   createdAt: '2022-12-11T20:16:13.797Z',
//   user: fakeUser(),
// });

export const fakeCartItem = (overrides: {}): TCartItem => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 4,
	product: fakeItem(),
	user: fakeUser({}),
	...overrides,
})

// // Fake LocalStorage
// class LocalStorageMock {
//   constructor() {
//     this.store = {};
//   }

//   clear() {
//     this.store = {};
//   }

//   getItem(key) {
//     return this.store[key] || null;
//   }

//   setItem(key, value) {
//     this.store[key] = value.toString();
//   }

//   removeItem(key) {
//     delete this.store[key];
//   }
// }

export function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	]
}

// export
// makePaginationMocksFor,
// LocalStorageMock,
// fakeItem,
// fakeUser,
// fakeCartItem,
// fakeOrder,
// fakeOrderItem,
