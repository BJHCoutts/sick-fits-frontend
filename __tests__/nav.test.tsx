import { render, screen } from '@testing-library/react'
import { CURRENT_USER_QUERY } from '../src/lib/graphQL/queries/currentUserQuery'
import { fakeCartItem, fakeUser } from '../src/lib/testUtils'
import { CartStateProvider } from '../src/lib/context/cartState'
import { MockedProvider } from '@apollo/client/testing'
import Nav from '../src/components/Nav'
import wait from 'waait'
import { act } from 'react-dom/test-utils'

const notSignedInMocks = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: { data: { authenticatedItem: null } },
	},
]

const signedInMocks = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: { data: { authenticatedItem: fakeUser({}) } },
	},
]

const signedInMocksWithCartItems = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: {
			data: {
				authenticatedItem: fakeUser({
					cart: [fakeCartItem({})],
				}),
			},
		},
	},
]

describe('<Nav />', () => {
	it('Renders a minimal nav when signed out', () => {
		const { container, debug } = render(
			<CartStateProvider>
				<MockedProvider mocks={notSignedInMocks}>
					<Nav />
				</MockedProvider>
			</CartStateProvider>
		)
		debug()
		expect(container).toHaveTextContent('Sign In / Up')
		expect(container).toMatchSnapshot()
		const link = screen.getByText('Sign In / Up')
		expect(link).toHaveAttribute('href', '/signin')
		debug()
	})

	it('renders a full nav when signed in', async () => {
		const { container, debug } = render(
			<CartStateProvider>
				<MockedProvider mocks={signedInMocks}>
					<Nav />
				</MockedProvider>
			</CartStateProvider>
		)

		await screen.findByText('Account')
		expect(container).toMatchSnapshot()
		expect(container).toHaveTextContent('Sign Out')
		expect(container).toHaveTextContent('My Cart')
	})

	it('renders the amount of items in the cart', async () => {
		const { container, debug } = render(
			<CartStateProvider>
				<MockedProvider mocks={signedInMocksWithCartItems}>
					<Nav />
				</MockedProvider>
			</CartStateProvider>
		)
		await wait(400)
		await screen.findByText('4')
	})
})
