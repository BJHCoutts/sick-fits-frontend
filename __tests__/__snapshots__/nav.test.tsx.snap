// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> Renders a minimal nav when signed out 1`] = `
<div>
  <nav
    class="sc-bdfCDU hTFCwp"
  >
    <ul>
      <li>
        <a
          href="/products"
        >
          Products
        </a>
      </li>
      <li>
        <a
          href="/signin"
        >
          Sign In / Up
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<Nav /> renders a full nav when signed in 1`] = `
<div>
  <nav
    class="sc-bdfCDU hTFCwp"
  >
    <ul>
      <li>
        <a
          href="/products"
        >
          Products
        </a>
      </li>
      <li>
        <a
          href="/sell"
        >
          Sell
        </a>
      </li>
      <li>
        <a
          href="/orders"
        >
          Orders
        </a>
      </li>
      <li>
        <a
          href="/account"
        >
          Account
        </a>
      </li>
      <li>
        <button
          type="button"
        >
          Sign Out
        </button>
      </li>
      <li>
        <button
          type="button"
        >
          My Cart
          <span
            class="sc-dlfmHC gBcOvK"
          >
            <div>
              <div
                class="sc-gsTDqH crGFiM count"
              >
                0
              </div>
            </div>
          </span>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
